In [47]:
import pandas as pd
import numpy as np
from sklearn.impute import KNNImputer
from sklearn.preprocessing import MinMaxScaler
import sys

In [2]:
data = pd.read_csv('data/pima-indians-diabetes.csv')
data.head()

6  148  72  35    0  33.6  0.627  50  1
0  1   85  66  29    0  26.6  0.351  31  0
1  8  183  64   0    0  23.3  0.672  32  1
2  1   89  66  23   94  28.1  0.167  21  0
3  0  137  40  35  168  43.1  2.288  33  1
4  5  116  74   0    0  25.6  0.201  30  0

In [3]:
data = pd.read_csv('data/pima-indians-diabetes.csv', names=['preg', 'plas', 'pres', 'skin', 'test', 'mass',' pedi', 'age', 'class'])
data.head()

preg  plas  pres  skin  test  mass   pedi  age  class
0     6   148    72    35     0  33.6  0.627   50      1
1     1    85    66    29     0  26.6  0.351   31      0
2     8   183    64     0     0  23.3  0.672   32      1
3     1    89    66    23    94  28.1  0.167   21      0
4     0   137    40    35   168  43.1  2.288   33      1

In [4]:
data.isnull()

preg   plas   pres   skin   test   mass   pedi    age  class
0    False  False  False  False  False  False  False  False  False
1    False  False  False  False  False  False  False  False  False
2    False  False  False  False  False  False  False  False  False
3    False  False  False  False  False  False  False  False  False
4    False  False  False  False  False  False  False  False  False
..     ...    ...    ...    ...    ...    ...    ...    ...    ...
763  False  False  False  False  False  False  False  False  False
764  False  False  False  False  False  False  False  False  False
765  False  False  False  False  False  False  False  False  False
766  False  False  False  False  False  False  False  False  False
767  False  False  False  False  False  False  False  False  False

[768 rows x 9 columns]

In [5]:
data.isnull().sum()

preg     0
plas     0
pres     0
skin     0
test     0
mass     0
 pedi    0
age      0
class    0
dtype: int64

In [6]:
data.isna()

preg   plas   pres   skin   test   mass   pedi    age  class
0    False  False  False  False  False  False  False  False  False
1    False  False  False  False  False  False  False  False  False
2    False  False  False  False  False  False  False  False  False
3    False  False  False  False  False  False  False  False  False
4    False  False  False  False  False  False  False  False  False
..     ...    ...    ...    ...    ...    ...    ...    ...    ...
763  False  False  False  False  False  False  False  False  False
764  False  False  False  False  False  False  False  False  False
765  False  False  False  False  False  False  False  False  False
766  False  False  False  False  False  False  False  False  False
767  False  False  False  False  False  False  False  False  False

[768 rows x 9 columns]

In [7]:
data.isna().sum()

preg     0
plas     0
pres     0
skin     0
test     0
mass     0
 pedi    0
age      0
class    0
dtype: int64

#### so our data hasn't any null or Nan vlaues apparently but will see... 

In [8]:
skines = data[data['skin'] == 0]
skines

preg  plas  pres  skin  test  mass   pedi  age  class
2       8   183    64     0     0  23.3  0.672   32      1
5       5   116    74     0     0  25.6  0.201   30      0
7      10   115     0     0     0  35.3  0.134   29      0
9       8   125    96     0     0   0.0  0.232   54      1
10      4   110    92     0     0  37.6  0.191   30      0
..    ...   ...   ...   ...   ...   ...    ...  ...    ...
757     0   123    72     0     0  36.3  0.258   52      1
758     1   106    76     0     0  37.5  0.197   26      0
759     6   190    92     0     0  35.5  0.278   66      1
762     9    89    62     0     0  22.5  0.142   33      0
766     1   126    60     0     0  30.1  0.349   47      1

[227 rows x 9 columns]

#### the fact is in this data in skin and test columns zero values means null for 227 rows
so we have several options : we can drop these columns or drop 227 rows or replace 0 values with mean

In [20]:
data['skin'] != 0

0       True
1       True
2      False
3       True
4       True
       ...  
763     True
764     True
765     True
766    False
767     True
Name: skin, Length: 768, dtype: bool

In [21]:
data[data['skin'] != 0]

preg  plas  pres  skin  test  mass   pedi  age  class
0       6   148    72    35     0  33.6  0.627   50      1
1       1    85    66    29     0  26.6  0.351   31      0
3       1    89    66    23    94  28.1  0.167   21      0
4       0   137    40    35   168  43.1  2.288   33      1
6       3    78    50    32    88  31.0  0.248   26      1
..    ...   ...   ...   ...   ...   ...    ...  ...    ...
761     9   170    74    31     0  44.0  0.403   43      1
763    10   101    76    48   180  32.9  0.171   63      0
764     2   122    70    27     0  36.8  0.340   27      0
765     5   121    72    23   112  26.2  0.245   30      0
767     1    93    70    31     0  30.4  0.315   23      0

[541 rows x 9 columns]

In [26]:
data[data['skin'] != 0]['skin']

0      35
1      29
3      23
4      35
6      32
       ..
761    31
763    48
764    27
765    23
767    31
Name: skin, Length: 541, dtype: int64

In [23]:
data[data['skin'] != 0]['skin'].mean()

29.153419593345657

In [27]:
skin_mean = data[data['skin'] != 0]['skin'].mean()

In [30]:
data = data.replace({'skin':0}, skin_mean)

In [31]:
data

preg  plas  pres      skin  test  mass   pedi  age  class
0       6   148    72  35.00000     0  33.6  0.627   50      1
1       1    85    66  29.00000     0  26.6  0.351   31      0
2       8   183    64  29.15342     0  23.3  0.672   32      1
3       1    89    66  23.00000    94  28.1  0.167   21      0
4       0   137    40  35.00000   168  43.1  2.288   33      1
..    ...   ...   ...       ...   ...   ...    ...  ...    ...
763    10   101    76  48.00000   180  32.9  0.171   63      0
764     2   122    70  27.00000     0  36.8  0.340   27      0
765     5   121    72  23.00000   112  26.2  0.245   30      0
766     1   126    60  29.15342     0  30.1  0.349   47      1
767     1    93    70  31.00000     0  30.4  0.315   23      0

[768 rows x 9 columns]

In [34]:
test_mean = data[data['test'] != 0]['test'].mean()
test_mean

155.5482233502538

In [35]:
data.replace({'test':0}, test_mean, inplace= True)

In [36]:
data

preg  plas  pres      skin        test  mass   pedi  age  class
0       6   148    72  35.00000  155.548223  33.6  0.627   50      1
1       1    85    66  29.00000  155.548223  26.6  0.351   31      0
2       8   183    64  29.15342  155.548223  23.3  0.672   32      1
3       1    89    66  23.00000   94.000000  28.1  0.167   21      0
4       0   137    40  35.00000  168.000000  43.1  2.288   33      1
..    ...   ...   ...       ...         ...   ...    ...  ...    ...
763    10   101    76  48.00000  180.000000  32.9  0.171   63      0
764     2   122    70  27.00000  155.548223  36.8  0.340   27      0
765     5   121    72  23.00000  112.000000  26.2  0.245   30      0
766     1   126    60  29.15342  155.548223  30.1  0.349   47      1
767     1    93    70  31.00000  155.548223  30.4  0.315   23      0

[768 rows x 9 columns]

#### an expert way to rpelace null values is this :

In [37]:
data = pd.read_csv('data/pima-indians-diabetes.csv', names=['preg', 'plas', 'pres', 'skin', 'test', 'mass',' pedi', 'age', 'class'])
data.head()

preg  plas  pres  skin  test  mass   pedi  age  class
0     6   148    72    35     0  33.6  0.627   50      1
1     1    85    66    29     0  26.6  0.351   31      0
2     8   183    64     0     0  23.3  0.672   32      1
3     1    89    66    23    94  28.1  0.167   21      0
4     0   137    40    35   168  43.1  2.288   33      1

In [38]:
data.replace({'skin':0, 'test':0}, np.nan, inplace=True)

In [39]:
data

preg  plas  pres  skin   test  mass   pedi  age  class
0       6   148    72  35.0    NaN  33.6  0.627   50      1
1       1    85    66  29.0    NaN  26.6  0.351   31      0
2       8   183    64   NaN    NaN  23.3  0.672   32      1
3       1    89    66  23.0   94.0  28.1  0.167   21      0
4       0   137    40  35.0  168.0  43.1  2.288   33      1
..    ...   ...   ...   ...    ...   ...    ...  ...    ...
763    10   101    76  48.0  180.0  32.9  0.171   63      0
764     2   122    70  27.0    NaN  36.8  0.340   27      0
765     5   121    72  23.0  112.0  26.2  0.245   30      0
766     1   126    60   NaN    NaN  30.1  0.349   47      1
767     1    93    70  31.0    NaN  30.4  0.315   23      0

[768 rows x 9 columns]

In [42]:
imputer  =KNNImputer(n_neighbors=3) # k = 3
data = imputer.fit_transform(data) # return ndarray of numpy
data

array([[  6.   , 148.   ,  72.   , ...,   0.627,  50.   ,   1.   ],
       [  1.   ,  85.   ,  66.   , ...,   0.351,  31.   ,   0.   ],
       [  8.   , 183.   ,  64.   , ...,   0.672,  32.   ,   1.   ],
       ...,
       [  5.   , 121.   ,  72.   , ...,   0.245,  30.   ,   0.   ],
       [  1.   , 126.   ,  60.   , ...,   0.349,  47.   ,   1.   ],
       [  1.   ,  93.   ,  70.   , ...,   0.315,  23.   ,   0.   ]])

In [43]:
type(data)

numpy.ndarray

In [45]:
np.set_printoptions(threshold=sys.maxsize) # to show all datas

In [46]:
data

array([[6.00000000e+00, 1.48000000e+02, 7.20000000e+01, 3.50000000e+01,
        1.25333333e+02, 3.36000000e+01, 6.27000000e-01, 5.00000000e+01,
        1.00000000e+00],
       [1.00000000e+00, 8.50000000e+01, 6.60000000e+01, 2.90000000e+01,
        6.66666667e+01, 2.66000000e+01, 3.51000000e-01, 3.10000000e+01,
        0.00000000e+00],
       [8.00000000e+00, 1.83000000e+02, 6.40000000e+01, 3.00000000e+01,
        1.95000000e+02, 2.33000000e+01, 6.72000000e-01, 3.20000000e+01,
        1.00000000e+00],
       [1.00000000e+00, 8.90000000e+01, 6.60000000e+01, 2.30000000e+01,
        9.40000000e+01, 2.81000000e+01, 1.67000000e-01, 2.10000000e+01,
        0.00000000e+00],
       [0.00000000e+00, 1.37000000e+02, 4.00000000e+01, 3.50000000e+01,
        1.68000000e+02, 4.31000000e+01, 2.28800000e+00, 3.30000000e+01,
        1.00000000e+00],
       [5.00000000e+00, 1.16000000e+02, 7.40000000e+01, 1.83333333e+01,
        1.09000000e+02, 2.56000000e+01, 2.01000000e-01, 3.00000000e+01,
        0.0

#### now we want to normalize data with min-max scaler

In [48]:
scaler = MinMaxScaler(feature_range=(0, 1))
data = scaler.fit_transform(data)

In [49]:
data

array([[0.35294118, 0.74371859, 0.59016393, 0.30434783, 0.1338141 ,
        0.50074516, 0.23441503, 0.48333333, 1.        ],
       [0.05882353, 0.42713568, 0.54098361, 0.23913043, 0.06330128,
        0.39642325, 0.11656704, 0.16666667, 0.        ],
       [0.47058824, 0.91959799, 0.52459016, 0.25      , 0.21754808,
        0.34724292, 0.25362938, 0.18333333, 1.        ],
       [0.05882353, 0.44723618, 0.54098361, 0.17391304, 0.09615385,
        0.41877794, 0.03800171, 0.        , 0.        ],
       [0.        , 0.68844221, 0.32786885, 0.30434783, 0.18509615,
        0.64232489, 0.94363792, 0.2       , 1.        ],
       [0.29411765, 0.58291457, 0.60655738, 0.12318841, 0.11418269,
        0.38152012, 0.05251921, 0.15      , 0.        ],
       [0.17647059, 0.3919598 , 0.40983607, 0.27173913, 0.08894231,
        0.46199702, 0.07258753, 0.08333333, 1.        ],
       [0.58823529, 0.57788945, 0.        , 0.25362319, 0.07451923,
        0.52608048, 0.02391119, 0.13333333, 0.        ],
